doc/style-guide-testing.md

```diff
@@ -0,0 +1,467 @@
+Writing High Quality Tests
+==========================
+
+The goal of a test is to ensure that some bit of code works as expected with valid or invalid input.  Testing failure paths is as important as testing successful paths.
+
+Tests also provide a way of proving that your interface doesn't change.  People will write code that use your module in a defined way, which should mirror a test.  When you refactor code for efficiency, to split into smaller modules or to use a different method for performing the same action your tests will guarantee that everything still works.
+
+Think of tests as a safety net for acrobats.  You'd rather be perfect and perform your art without ever needing tests, but they are there to help you when a mistake is made.
+
+
+Guidelines
+----------
+
+The [style guide](style-guide.md) lists many useful tips and various requirements when writing tests.  Only a few of those are listed here for further explanation.
+
+
+### Eliminate Duplicated Setup
+
+When the code has two separate actions that it performs then you want to have the tests split into two separate `it()` calls.  These would require the same setup to achieve an identical scenario.  Move that into a `beforeEach()`.  When you have a complicated piece of code that has numerous logical branches, again you will want to split the setup into multiple `beforeEach()` calls, one for each branch.
+
+There's examples in this document to help explain this process.
+
+
+### Make it easy to expand
+
+Let's say that your code currently does only one thing.  Perhaps it just trims whitespace.  You'd see no reason to have a `beforeEach()` that would set up some service and then a separate `it()` to test if the whitespace was removed.  You could put all of the setup and the testing together in an `it()`.
+
+It is recommended to not do this.  The `it()` is considered the "test" portion of the code and `beforeEach()` is reserved for setup.  Think about what you would have to do in order to prove a bug was fixed properly.  If the setup was combined with the test, then you'd either have to duplicate the setup in the second test (already you know that's bad) or you'll have to refactor the test to have a `beforeEach()` and two `it()` calls.  Since the latter option is the one we'd like to see, try to make all setup happen in a `beforeEach()` even if there is only one test.  It makes the tests easier to expand.
+
+
+### Test Behavior
+
+We want to write tests to not only confirm there are no defects, but also ones that could mimic how the software is intended to be used.  By testing behavior, we are shifting our viewpoint to be from the user's point of view instead of being overly concerned with the internals of the software.  In essence, I want to make sure your code performs a specific action to produce the desired result.  I do not want to guarantee that the code does every specific step in exactly the right order.
```

Comment:
I think we should stay with the `we` context. Switching to `I` in the middle seems a bit odd. 

    In essence, we want to make sure the code performs a specific action to produce the desired result.  We do not want to guarantee the code does every specific step in exactly the right order.